### PR TITLE
fix(pricing): recalc uses historical price snapshot via attempt.ModelPriceID

### PR DESCRIPTION
**Pull request by Bowl42**
## Summary

之前 \`RecalcAttemptUpdate\` 用模型名调 \`Calculator.Calculate\`,实际上取的是\`pricesByKey\` 里的**当前价**,等于把"重算"语义偷换成"按今天的价回填历史 attempt"。这违背了重算的真实意图——\`token\` 后补 / 流式结尾丢 metrics / 手工 backfill 等场景需要的是"按**当时价** × 最新 token × 历史 multiplier"重算。

随着 #574 让价格表版本化(\`Update\` 软删旧行 + INSERT 新行,旧行物理保留),\`attempt.ModelPriceID\` 已经是一个稳定的"当时价"句柄,可以接进重算路径。

## Changes

1. **\`ModelPriceRepository.GetByIDIncludingDeleted(id)\`** — 绕过 \`deleted_at = 0\` 过滤,专给 attempt 历史快照反查用。与 #574 review 中和 maintainer 达成的约定一致(独立方法,不削弱 \`GetByID\` 的 admin \"Delete → 404\" 语义)。

2. **\`pricing.Calculator\`**:
    - 新字段 \`historicalLookup HistoricalLookup\` + \`SetHistoricalLookup(fn)\` 注入点
    - 新方法 \`CalculateByPriceID(id, metrics, multiplier) (CostResult, bool)\` — 先查内存 \`pricesByID\` 缓存,未命中走 \`historicalLookup\` 懒加载并双检入缓存(避免重算大批 attempt 打爆 DB);返回 \`ok=false\` 时上层回退
    - 抽出 \`applyPrice\` 共享 cost × multiplier 公式

3. **\`RecalcAttemptUpdate\` 定价决策**:
    \`\`\`
    attempt.ModelPriceID > 0 → CalculateByPriceID(id, ...)            ← 当时价
    miss / ID == 0           → fallback Calculate(model, ...)         ← 当前价兜底
    \`\`\`

4. **\`core/database.go\` 启动 wiring**:\`Calculator.SetHistoricalLookup(repo.GetByIDIncludingDeleted)\`。

## 升级影响

- **无 schema 变化、无 migration**
- 现有 \`attempt.ModelPriceID\` 字段早就有,此前虽然写入但不被读;现在读它走当时价
- v0.13.82(本次基线)所有现存 attempt 的 \`ModelPriceID\` 都是 seed 时分配的当前 ID,价格从未版本化分裂过,所以本次部署不会改变任何已有数值
- 第一次 admin 改价之后,**那时**起新写入的 attempt 拿到新版本 ID;之后任何重算都会自动用当时的价格快照,**不会因为新的价格而把历史 cost 改写**
- 反向用例(故意按新价回填):本 PR 不动这条路径,如果未来需要可以另开独立 admin 操作显式触发

## Test plan

- [x] \`go test ./internal/pricing/\` — 新增 7 个子测试通过(by-ID 取快照不被同名当前价覆盖、lazy-load 缓存只回 DB 一次、ID=0 / 无 lookup / lookup 返回 nil 三种 fallback 路径、Recalc 端到端验证)
- [x] \`go test ./internal/service/ ./internal/repository/sqlite/ ./tests/e2e/\` 全包无回归
- [x] \`go build ./internal/... ./cmd/...\` 干净
- [ ] 部署后:在 admin UI 改一次 sonnet 价格,生成新 attempt,然后改回原价,触发"重算成本"——确认 cost 保持当时价不漂移

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * 支持按历史价格快照重算已有尝试的成本，并能按 ID 读取包含已软删除的历史价格记录。

* **Bug Fixes**
  * 当历史快照可用时优先使用其单价；历史查表失败时保留原始成本并避免不安全回退。
  * 在历史快照缺失或未命中时，自动降级为按模型当前价格重算。

* **Tests**
  * 增加覆盖历史查表与重算契约的自动化测试并改进全局重置，防止测试间污染。

[Review Change Stack](https://app.coderabbit.ai/change-stack/awsl-project/maxx/pull/575?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by Bowl42** [comment]
@coderabbitai review

File: internal/pricing/calculator.go [hunk 1]
```diff
@@ -82,13 +91,53 @@ func (c *Calculator) GetModelPrice(model string) *domain.ModelPrice {
...
+	lookup := c.historicalLookup
+	c.mu.RUnlock()
+
+	if lookup == nil {
+		return nil
+	}
+	p, err := lookup(id)
+	if err != nil || p == nil {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (awsl233777) [reply internal/pricing/calculator.go:135]: I agree this should block merge until fixed.

This path is part of the pricing audit chain, so a transient `historicalLookup` / DB error must not be treated the same as a legitimate “historical price not found” case. Falling back to `Calculate(model, ...)` on a lookup error can silently rewrite a historical attempt with the current price and a different `model_price_id`, which is exactly the contract this PR is trying to protect.

Please distinguish the cases explicitly:

- `p == nil` / not found: fallback to current-price calculation is acceptable for old or corrupted records.
- `err != nil`: do not fallback to current price. Propagate/return the error, or at minimum leave the attempt unchanged and make the failure observable.

Please also add a regression test where `historicalLookup` returns a non-nil error and assert that it does not update cost/model_price_id via the current-price fallback.

**Comment by Bowl42** [comment]
@awsl233777 @coderabbitai — addressed in [`ccefa40`](https://github.com/awsl-project/maxx/pull/575/commits/ccefa40) (cherry-picked from PR #576) + [`49aac87`](https://github.com/awsl-project/maxx/pull/575/commits/49aac87).

**1. silent-error-swallow (must-fix, both reviewers)** — fixed exactly as suggested:
- `resolvePriceByID` → `(*ModelPrice, error)`: propagates lookup error; only `p == nil` (legitimate "not found") returns `(nil, nil)`.
- `CalculateByPriceID` → `(CostResult, bool, error)`: error is its own return branch, distinct from the genuine-miss `(_, false, nil)`.
- `RecalcAttemptUpdate`: on `err != nil`, logs the error and returns `(currentCost, AttemptCostUpdate{}, false)` — attempt unchanged, no fallback to current-price calculation. The audit chain is preserved on transient DB errors.

**2. Regression tests** for both reviewer asks:
- `TestRecalcFromAttempt_DoesNotFallBackOnLookupError` (`attempt_test.go`) — `historicalLookup` returns a synthetic error, assert `changed=false` and that `newCost` / `update` are unchanged.
- new `lookup returns error` subtest in `TestCalculator_CalculateByPriceID_NoLookupOrUnknownID` — asserts the error is propagated through `CalculateByPriceID` itself.
- `TestGetByIDIncludingDeleted_ReturnsSoftDeletedRow` (`model_price_test.go`, my own addition from internal review) — locks the positive contract: after `Delete` (soft-delete), `GetByID` 404s but `GetByIDIncludingDeleted` still returns the row.

**On the internal Claude review's "cross-model sanity check" finding** (CalculateByPriceID doesn't verify the looked-up row's ModelID matches `attempt.ResponseModel`): leaving as follow-up. The corruption scenario is data-state, not a contract gap introduced by this PR; the same hazard exists today for the live pricing path. Happy to handle as a separate PR with explicit ModelID-mismatch fallback if you'd like.

CI re-running on `49aac87`.

**Comment by awsldev** [comment]
Re-check after `ccefa40` / `49aac87`:

The original audit-chain blocker is fixed correctly now: lookup errors are propagated, `RecalcAttemptUpdate` preserves the existing attempt on error, and the regression tests cover the no-fallback path. I also locally ran:

```text
go test ./cmd/... ./internal/... ./tests/...
```

and it passed.

One remaining thing before approval: please address CodeRabbit’s latest nit in `TestCalculator_CalculateByPriceID_LazyLoadCaches`. The loop currently discards `ok` / `err`; that makes the cache test weaker than it should be because it can still pass if `CalculateByPriceID` regresses to returning `ok=false` or an error while only the lookup call count stays at 1.

This should be a tiny test-only change: capture `_, ok, err := calc.CalculateByPriceID(...)` inside the loop and assert `err == nil` and `ok == true` on each iteration.

For the cross-model sanity check deferral: I’m okay with that staying out of this PR, but please track it as a follow-up because it is still adjacent to the same pricing/audit surface.

**Comment by Bowl42** [comment]
@awsl233777 — fixed in [`78e6e47`](https://github.com/awsl-project/maxx/pull/575/commits/78e6e47). The loop now captures `ok, err` and asserts `err == nil && ok == true` on each iteration, so a future regression that flips `CalculateByPriceID` to return `ok=false` (e.g. a stale cache invalidation bug) fails the test even if `calls == 1` stays accurate.

On the **cross-model sanity check** follow-up: I'll open a tracking issue right after this PR lands so it doesn't get lost. The scope is small but it adds a parameter to `CalculateByPriceID` (`wantModelID string`) plus a stub-repo update across the test fakes, so cleaner as its own PR.

@coderabbitai — please go ahead and open that follow-up issue you offered. Reference: PR #575 review thread on cross-model ModelID guard in `CalculateByPriceID`.

**Review by awsl233777 — APPROVED** [review]
Approved after checking the current gate: all required checks are green, CodeRabbit has no actionable comments, the prior audit-chain blocker was fixed, and the remaining cross-model ModelID guard is tracked separately in issue #577.
